### PR TITLE
IPI E2E CI steps, chains and workflow for Nutanix clusters

### PR DESCRIPTION
**Pull request by adiantum**
PR contains steps, chains and workflow for E2E CI for Nutanix clusters.

Notes:
 - DNS records steps implementation based on vSphere steps
 - It's assumed that networks in CI Nutanix cluster have the same name like LEASED_RESOURCE
 - It's assumed that networks in CI Nutanix cluster have subnet network address (L3 network is managed), otherwise environment variables for VIPs (API, INGRESS) should be defined
 - Is's assumed that Prism Central have only one Prism Element

### CODE REVIEW

**Comment by rvanderp3** [comment]
@adiantum you'll need to configure a `nutanix` cluster profile in https://github.com/openshift/ci-tools/ before this PR can proceed.  I didn't see anything immediately concerning, but will give this a deeper look once the cluster profile is in place.

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+set -o errexit
+set -o pipefail
+
+echo "${BASE_DOMAIN:?BASE_DOMAIN env variable should be defined}" > "${SHARED_DIR}"/basedomain.txt
+
+cluster_name=${NAMESPACE}-${JOB_NAME_HASH}
+base_domain=$(<"${SHARED_DIR}"/basedomain.txt)
+cluster_domain="${cluster_name}.${base_domain}"
+
+export AWS_DEFAULT_REGION=us-west-2  # TODO: Derive this?
```
Comment (JoelSpeed) [inline ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:13]: Have you tested this on any other regions?

Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:13]: > Have you tested this on any other regions?
Yes, it's working. Just also tested with us-west-1.

Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:13]: > Have you tested this on any other regions?
Yes, it's working. Just also tested with us-west-1.

**Review by elmiko — COMMENTED** [review]
this generally looks good to me, +1

**Review by elmiko — COMMENTED** [review]
this looks good to me, i'd like to get confirmation on the `cluster` field from someone who knows a little better than me ;)
/lgtm

**Review by elmiko — COMMENTED** [review]
i think this is still looking good. will we need to get the MAO changes in place for this to pass the rehearse tests? (cc @JoelSpeed )

**Comment by JoelSpeed** [comment]
As this is adding presubmits to the installer, someone from that team should have the final ACK IMO, CC @patrickdillon

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    "ResourceRecordSet": {
+      "Name": "api.${cluster_domain}.",
+      "Type": "A",
+      "TTL": 60,
+      "ResourceRecords": [{"Value": "${API_VIP}"}]
+      }
+    },{
+    "Action": "UPSERT",
+    "ResourceRecordSet": {
+      "Name": "api-int.${cluster_domain}.",
```
Comment (jcpowermac) [inline ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:61]: api-int should really not be here

This was added to vSphere for the windows folks but only as a workaround. The api-int A record should be supplied by coredns static pod

File: ci-operator/config/openshift/installer/openshift-installer-release-4.11.yaml [hunk 3]
```diff
@@ -484,6 +484,11 @@ tests:
   steps:
     cluster_profile: ibmcloud
     workflow: openshift-e2e-ibmcloud
+- as: e2e-nutanix
+  optional: true
```
Comment (staebler) [inline ci-operator/config/openshift/installer/openshift-installer-release-4.11.yaml:488]:
```suggestion
  optional: true
  run_if_changed: nutanix
```

File: ci-operator/jobs/openshift/installer/openshift-installer-release-4.11-presubmits.yaml [hunk 4]
```diff
@@ -148,7 +148,7 @@ presubmits:
           secretName: result-aggregator
     trigger: (?m)^/test( | .* )e2e-aws,?($|\s.*)
   - agent: kubernetes
-    always_run: false
+    always_run: true
```
Comment (staebler) [inline ci-operator/jobs/openshift/installer/openshift-installer-release-4.11-presubmits.yaml]: Most of the changes in this file should not be made.

File: ci-operator/config/openshift/installer/openshift-installer-master.yaml [hunk 5]
```diff
@@ -487,3 +487,10 @@ zz_generated_metadata:
   branch: master
   org: openshift
   repo: installer
+- as: e2e-nutanix
+  optional: true
```
Comment (staebler) [inline ci-operator/config/openshift/installer/openshift-installer-master.yaml:487]:
```suggestion
  optional: true
  run_if_changed: nutanix
```

File: ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml [hunk 6]
```diff
@@ -4186,3 +4186,76 @@ presubmits:
           requests:
             cpu: 500m
     trigger: (?m)^/test yaml-lint
+  - agent: kubernetes
+    always_run: true
```
Comment (staebler) [inline ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml]:
```suggestion
    always_run: false
```

File: ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml [hunk 7]
```diff
@@ -4186,3 +4186,76 @@ presubmits:
...
+    context: ci/prow/e2e-nutanix
+    decorate: true
+    labels:
+      ci-operator.openshift.io/cloud: nutanix
+      ci-operator.openshift.io/cloud-cluster-profile: nutanix
+      ci.openshift.io/generator: prowgen
+      pj-rehearse.openshift.io/can-be-rehearsed: "true"
+    name: pull-ci-openshift-installer-e2e-nutanix
+    rerun_command: /test e2e-nutanix
+    skip_if_only_changed: (^docs/)|((^|/)OWNERS(_ALIASES)?$)|((^|/)[A-Z]+\.md$)|(^(data/.*|pkg/.*|images)/(alibabacloud|aws|baremetal|gcp|ibmcloud|libvirt|openstack|ovirt|vsphere|azure)/)
```
Comment (staebler) [inline ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml]: Use a `run_if_changed` sourced from the config file.

File: ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml [hunk 8]
```diff
@@ -4186,3 +4186,76 @@ presubmits:
...
+        - name: cluster-profile
+          secret:
+            secretName: cluster-secrets-nutanix
+        - name: pull-secret
+          secret:
+            secretName: registry-pull-credentials
+        - name: result-aggregator
+          secret:
+            secretName: result-aggregator
+    trigger: (?m)^/test( | .* )e2e-aws,?($|\s.*)
```
Comment (staebler) [inline ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml]: Was this section copied? There is a make target to generate the file.

**Review by staebler — COMMENTED** [review]
Please fix up the commits so that they do not include bad changes that are fixed in later commits.

File: ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml (hunk 8 above)
Comment (adiantum) [reply ci-operator/jobs/openshift/installer/openshift-installer-master-presubmits.yaml]: Yes, it was copied. That time we have no access to registry with images required for `make jobs`

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh (hunk 2 above)
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:61]: We've checked the install process without the record. Master nodes even can not successfully start.
Looks like we need the workaround as well like vSphere
Screenshot 2022-04-14 at 17 12 43
.

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+base_domain=$(<"${SHARED_DIR}"/basedomain.txt)
+cluster_domain="${cluster_name}.${base_domain}"
+
+export AWS_DEFAULT_REGION=us-west-2  # TODO: Derive this?
+export AWS_SHARED_CREDENTIALS_FILE=/var/run/vault/nutanix/.awscred
+export AWS_MAX_ATTEMPTS=50
+export AWS_RETRY_MODE=adaptive
+export HOME=/tmp
+
+if ! command -v aws &> /dev/null
```
[the remark this answers is not in the document]
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:19]: We're using `tools` image. I'm not sure if it contains `aws` util, but the same code working in vSphere steps.

File: ci-operator/step-registry/ipi/conf/nutanix/manual-creds/ipi-conf-nutanix-manual-creds-commands.sh [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+kind: Secret
+metadata:
+   name: nutanix-credentials
+   namespace: openshift-machine-api
+type: Opaque
+data:
+  NUTANIX_ENDPOINT: ${NTNX_ENDPOINT}
+  NUTANIX_PORT: ${NTNX_PORT}
+  NUTANIX_USER: ${NTNX_USER}
+  NUTANIX_PASSWORD: ${NTNX_PASSWORD}
```
[the remark this answers is not in the document]
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/manual-creds/ipi-conf-nutanix-manual-creds-commands.sh:26]: Actually should not. The logic in this part is the same as in install-config.yaml template.

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh (hunk 2 above)
Comment (jcpowermac) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:61]: That is a bug then

see:
https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/node.go#L32-L44

File: ci-operator/config/openshift/installer/openshift-installer-master.yaml (hunk 5 above)
Comment (adiantum) [reply ci-operator/config/openshift/installer/openshift-installer-master.yaml:487]: Done

File: ci-operator/config/openshift/installer/openshift-installer-release-4.11.yaml (hunk 3 above)
Comment (adiantum) [reply ci-operator/config/openshift/installer/openshift-installer-release-4.11.yaml:488]: Done

File: ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh (hunk 2 above)
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/dns/ipi-conf-nutanix-dns-commands.sh:60]: Thank you. We'll remove it.

**Comment by jcpowermac** [comment]
@adiantum missing a nutanix section for:
https://github.com/openshift/release/blob/master/core-services/ci-secret-bootstrap/_config.yaml
thats why we are getting `Pod pending timeout`

File: ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-chain.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+chain:
+  as: ipi-conf-nutanix
+  steps:
+  - ref: ipi-conf-nutanix-context
+  - ref: ipi-conf-nutanix-dns
+  - ref: ipi-conf
+  - ref: ipi-conf-nutanix
+  - ref: ipi-conf-nutanix-manual-creds
```
Comment (jsafrane) [inline ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-chain.yaml:8]: Some StatefulSet tests in `openshift-e2e-test` require storage + default storage class to be available. I recommend to install also your CSI driver as a new step openshift-e2e-test, say `ipi-conf-nutanix-csi`. Just put all manifests that OLM + your CSI driver needs into `$SHARED_DIR/` with names matching `manifest_*.yaml` and CI will add them to installation manifests and installer will create them. During installation, OLM should find them and install the operator.

Alternatively, you can create a new step *after* installation but before running any e2e tests in `openshift-e2e-nutanix-workflow.yaml`. You have a live cluster available at that time and you can do anything you want to install + configure the CSI driver.

Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-chain.yaml:8]: Thank you, @jsafrane!
We're working on CSI install part and planned to make it in separate PR.

File: ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+source "${SHARED_DIR}/nutanix_context.sh"
+
+echo "$(date -u --rfc-3339=seconds) - Adding platform data to install-config.yaml"
+
+# Populate install-config with Nutanix specifics
+cat >> "${CONFIG}" << EOF
+baseDomain: ${BASE_DOMAIN}
+credentialsMode: Manual
+platform:
```
[the remark this answers is not in the document]
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh:23]: I think CI steps should be changed when the installer PR will be approved or merged.

File: ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-chain.yaml (hunk 11 above)
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-chain.yaml:8]: CSI step added.

File: ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (adiantum) [reply ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh:23]: Fixed

File: ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+        address: ${NUTANIX_HOST}
+        port: ${NUTANIX_PORT}
+      password: ${NUTANIX_PASSWORD}
+      username: ${NUTANIX_USERNAME}
+    prismElements:
+    - endpoint:
+        address: ${PE_HOST}
+        port: ${PE_PORT}
+      uuid: ${PE_UUID}
+    subnetUUID:
```
Comment (yanhua121) [inline ci-operator/step-registry/ipi/conf/nutanix/ipi-conf-nutanix-commands.sh:38]: This should be "subnetUUIDs"

**Review by hongkailiu — COMMENTED** [review]
Just for the reference in the future.
The job is pinged on b01 because it uses VPN which is currently work only on b01.

Please remove `cluster: ` in the tests when VPN works on all clusters.
It will enable the failover of cluster.

**Review by hongkailiu — COMMENTED** [review]
/approve

**Pull request merged**